File: cOS/test/test_cOS.js


// Vendor Modules
/////////////////////////
var _ = require('lodash')
var expect = require('expect.js')
var describe = global.describe
var it = global.it
var async = require('async')
// var fs = require('fs')
// var path = require('path')

// Our Modules
/////////////////////////

describe('cOS', function() {

	var cOS
	var searchPaths, extension

	it('should load', function() {
		cOS = require('../cOS/cOS')
	})

	it ('should getPathInfo', function() {
		var options = {
			root: 'test'
		}
		var fileInfo = cOS.getPathInfo('test/test-cOS/four.js', options)

		var expectedInfo = {
			basename: 'four.js',
			extension: 'js',
			name: 'four',
			dirname: 'test/test-cOS/',
			path: 'test/test-cOS/four.js',
			root: 'test/',
			relativeDirname: './test-cOS/',
			relativePath: './test-cOS/four.js',
			filebase: 'test/test-cOS/four'
		}
		expect(fileInfo).to.eql(expectedInfo)
	})

	it('should ensureEndingSlash', function() {
		expect(cOS.ensureEndingSlash('path/')).to.be('path/')
		expect(cOS.ensureEndingSlash('path')).to.be('path/')
		expect(cOS.ensureEndingSlash('path\\')).to.be('path/')
	})

	it('should normalizeDir', function() {
		expect(cOS.normalizeDir('some/path/')).to.be('some/path/')
		expect(cOS.normalizeDir('some/path')).to.be('some/path/')
		expect(cOS.normalizeDir('some\\path\\')).to.be('some/path/')
		expect(cOS.normalizeDir('some\\path')).to.be('some/path/')
		expect(cOS.normalizeDir('/some/path/')).to.be('/some/path/')
		expect(cOS.normalizeDir('/some/path')).to.be('/some/path/')
		expect(cOS.normalizeDir('\\some\\path\\')).to.be('/some/path/')
		expect(cOS.normalizeDir('\\some\\path')).to.be('/some/path/')
	})

	it('should normalizePath', function() {
		expect(cOS.normalizePath('some/path/file.js')).to.be('some/path/file.js')
		expect(cOS.normalizePath('some\\path\\file.js')).to.be('some/path/file.js')
		expect(cOS.normalizePath('/some/path/file.js')).to.be('some/path/file.js')
		expect(cOS.normalizePath('/some\\path\\file.js')).to.be('some/path/file.js')
	})

	it('should join', function() {
		expect(cOS.join('some/path/','file.js')).to.be('some/path/file.js')
		expect(cOS.join('some/path\\','file.js')).to.be('some/path/file.js')
		expect(cOS.join('/some/','/path/file.js')).to.be('/some/path/file.js')
		expect(cOS.join('some\\','\\path\\file.js')).to.be('some/path/file.js')
	})

	it('should collectFiles', function(done) {
		searchPaths = ['test/test-cOS']
		extension = 'mustache'
		function cb(err, files)
		{
			if (err) throw err
			expect(files).to.be.a('object')
			expect(files.length).to.be(3)
			var names = _.map(files, 'basename')
			expect(names).to.eql(['one.mustache','two.mustache','three.mustache'])
			done()
		}
		cOS.collectFiles(searchPaths, extension, cb)
	})

	it('should collectFiles with exclusions', function(done) {
		searchPaths = ['test/test-cOS']
		extension = 'mustache'
		function cb(err, files)
		{
			if (err) throw err

			expect(files).to.be.a('object')
			expect(files.length).to.be(2)
			var names = _.map(files, 'basename')
			expect(names).to.eql(['one.mustache', 'three.mustache'])
			done()
		}
		cOS.collectFiles(searchPaths, extension, {exclude: ['two']}, cb)
	})

	it('should collectFiles searchPaths and extensions', function(done) {
		searchPaths = ['test/test-cOS/one/two/','test/test-cOS/three']
		extension = ['mustache','styl']
		function cb(err, files)
		{
			if (err) throw err
			expect(files).to.be.a('object')
			expect(files.length).to.be(3)
			var names = _.map(files, 'basename')
			expect(names).to.contain('two.mustache','three.mustache','three.styl')
			done()
		}
		cOS.collectFiles(searchPaths, extension, cb)
	})

	it('should get the cwd', function() {
		console.log(cOS.cwd())
	})

	it ('should upADir', function(){
		var upDir = cOS.upADir('c:/test/sub')
		expect(upDir).to.be('c:/test/')
		upDir = cOS.upADir('c:/')
		expect(upDir).to.be('c:/')
		upDir = cOS.upADir('etc/')
		expect(upDir).to.be('etc/')

		var filename = 'r:/Aeroplane/Final_Renders/AER_Video/EXR_Linear/AER_Airplane_020_v004/AER_Airplane_020_v004.%04d.exr'
		expect(cOS.upADir(cOS.upADir(cOS.upADir(filename)))).to.be('r:/Aeroplane/Final_Renders/')

		filename = 'r:/Aeroplane/Final_Renders/AER_Video/EXR_Linear/AER_Airplane_020_v004/'
		expect(cOS.upADir(cOS.upADir(cOS.upADir(filename)))).to.be('r:/Aeroplane/Final_Renders/')

		filename = 'r:/Aeroplane/Final_Renders/AER_Video/EXR_Linear/AER_Airplane_020_v004'
		expect(cOS.upADir(cOS.upADir(cOS.upADir(filename)))).to.be('r:/Aeroplane/')

	})

	it('should collectFilesSync', function(done) {
		var files = cOS.collectFilesSync(__dirname)
		expect(files.length).to.be.greaterThan(0)
		done()
	})

	it ('should runCommand', function(done) {
		cOS.runCommand('ls', function(err, out, exitCode)
		{
			console.log('out:', out)
			console.log('err:', err)
			expect(out).to.contain('LICENSE\nREADME.md')
			expect(err).to.be('')
			expect(exitCode).to.be(0)
			done()
		})
	})

	it ('should getGlobalModulesDir', function(done) {
		this.timeout = 3000
		cOS.getGlobalModulesDir(function(err, path)
		{
			console.log('"' + path + '"')
			expect(path.length).to.be.ok()
			expect(err).to.be(false)
			done()
		})
	})

	it ('should unixPath', function(done) {
		expect(cOS.unixPath('\\path//to\\\\file')).to.be('/path/to/file')
		expect(cOS.unixPath('//\\path//to\\\\file\\\\\\')).to.be('/path/to/file/')
		expect(cOS.unixPath('\\\\\\\\//path\\to///\\/\\\\file\\/\\/\\\\//\\\\//')).to.be('/path/to/file/')
		done()
	})

	it ('should fileExtension', function(done) {
		expect(cOS.getExtension('test.txt')).to.be('txt')
		expect(cOS.getExtension('/path/to/file.html')).to.be('html')
		expect(cOS.getExtension('/path/to/file/with/no/extension')).to.be('')
		done()
	})

	it ('should removeExtension', function(done) {
		expect(cOS.removeExtension('text.txt')).to.be('text')
		expect(cOS.removeExtension('/path/to/file.html')).to.be('/path/to/file')
		expect(cOS.removeExtension('/path/to/file/with/no/extension')).to.be('/path/to/file/with/no/extension')
		done()
	})

	it ('should getVersion', function(done) {
		expect(cOS.getVersion('test_v001.txt')).to.be(1)
		expect(cOS.getVersion('test_v100')).to.be(100)
		expect(cOS.getVersion('test_v0421894.txt')).to.be(421894)
		expect(cOS.getVersion('test_v001.txt')).to.be(1)
		expect(cOS.getVersion('test/v012/filev012')).to.be(12)
		done()
	})

	it ('should incrementVersion', function(done) {
		expect(cOS.incrementVersion('filev002')).to.be('filev003')
		expect(cOS.incrementVersion('filev009')).to.be('filev010')
		expect(cOS.incrementVersion('/path/v002/filev002')).to.be('/path/v003/filev003')
		done()
	})

	it ('should padLeft', function(done) {
		expect(cOS.padLeft('16', '0', 4)).to.be('0016')
		expect(cOS.padLeft('142', '0', 3)).to.be('142')
		expect(cOS.padLeft('1717', '0', 2)).to.be('1717')
		done()
	})

	it ('should getFrameRange', function(done) {
		cOS.makeDirSync('seq')
		var funcs = _.map(_.range(10), function(i)
		{
			return function(callback)
			{
				cOS.runCommand(
					'touch',
					'seq/name_v001.' + cOS.padLeft(String(i + 1510), '0', 4) + '.jpg',
					callback)
			}
		})
		funcs.push(function(callback)
		{
			var info = cOS.getFrameRange('seq/name_v001.%04d.jpg')
			expect(info.min).to.be(1510)
			expect(info.max).to.be(1519)
			callback()
		})
		funcs.push(function(callback)
		{
			cOS.removeDirSync('seq')
			callback()
			// cOS.runCommand('rm', ['-rf, seq'], callback)
		})
		async.series(funcs, done)
		// for (var i = 0; i < 10; i+=1)
		// 	cOS.runCommand('touch', 'seq/name_v001.' + cOS.padLeft(String(i + 1510), '0', 4) + '.jpg')
		// var info = cOS.getFrameRange('seq/name_v001.%04d.jpg')
		// console.log('info:', info)
		// expect(info.min).to.be(1510)
		// expect(info.max).to.be(1519)
		// done()
	})
})

File: cOS/test/test_cOS.py

import os
import subprocess

import arkInit
arkInit.init()

import tryout
import cOS


class test(tryout.TestSuite):
	title = 'test/cOS.py'

	def setUp(self):
		os.system('rm -rf sandbox')
		os.mkdir('sandbox')
		open('sandbox/file_v001.mb', 'w')
		open('sandbox/file.mb', 'w')
		os.mkdir('sandbox/sandboxSubdir')
		open('sandbox/sandboxSubdir/file1.txt', 'w')
		os.mkdir('sandbox/testdir1')
		os.mkdir('sandbox/testdir2')
		open('sandbox/testdir1/file1', 'w')
		open('sandbox/testdir1/file2', 'w')
		open('sandbox/testdir1/file3', 'w')
		open('sandbox/testdir2/file1', 'w')
		os.mkdir('sandbox/seq')
		os.mkdir('sandbox/emptyDir')
		for i in range(10):
			open('sandbox/seq/frame.%04d.exr' % (i + 1510), 'w')

		open('sandbox/seq/newFrame.0001.exr', 'w')

	def tearDown(self):
		os.system('rm -rf sandbox')

	def makeDir(self):
		cOS.makeDir('testDir')
		self.assertTrue(os.path.isdir('testDir'))
		os.system('rmdir testDir')

	def getExtension(self):
		ext = cOS.getExtension('file_v001.mb')
		self.assertEqual(ext, 'mb')

	def getVersion(self):
		ver = cOS.getVersion('sandbox/file_v001.mb')
		self.assertEqual(ver, 1)

		ver = cOS.getVersion('654')
		self.assertEqual(ver, 654)

		ver = cOS.getVersion(27)
		self.assertEqual(ver, 27)


	def getVersionError(self):
		ver = cOS.getVersion('sandbox/file.mb')
		self.assertEqual(ver, 0)

	def incrementVersion(self):
		ver = cOS.incrementVersion('sandbox/file_v001.mb')
		self.assertEqual(cOS.getVersion(ver), 2)

	def getDirName(self):
		dirname = cOS.getDirName('sandbox/file.mb')
		self.assertEqual(dirname, 'sandbox/')

	def emptyDir(self):
		cOS.emptyDir('sandbox/')
		self.assertTrue(not subprocess.check_output(['ls', 'sandbox']).split())

	def getPathInfo(self):
		options = {'root': 'test'}
		info = cOS.getPathInfo('test/test-cOS/four.js', options)

		self.assertEqual(info['basename'], 'four.js')
		self.assertEqual(info['extension'], 'js')
		self.assertEqual(info['name'], 'four')
		self.assertEqual(info['dirname'], 'test/test-cOS/')
		self.assertEqual(info['path'], 'test/test-cOS/four.js')
		self.assertEqual(info['root'], 'test/')
		self.assertEqual(info['relativeDirname'], './test-cOS/')
		self.assertEqual(info['relativePath'], './test-cOS/four.js')
		self.assertEqual(info['filebase'], 'test/test-cOS/four')

	def removeExtension(self):
		stripped = cOS.removeExtension('sandbox/file_v001.mb')
		self.assertEqual(stripped, 'sandbox/file_v001')

	def removeExtensionNoExtension(self):
		stripped = cOS.removeExtension('path/to/file')
		self.assertEqual(stripped, 'path/to/file')

	def unixPath(self):
		prepped = cOS.unixPath('\\sandbox\\file_v001.mb')
		self.assertEqual(prepped, '/sandbox/file_v001.mb')

	def ensureEndingSlash(self):
		normalized = cOS.ensureEndingSlash('path/to/dir')
		self.assertEqual(normalized, 'path/to/dir/')
		normalized = cOS.ensureEndingSlash('http://some/url')
		self.assertEqual(normalized, 'http://some/url/')

	def duplicateDir(self):
		cOS.duplicateDir('sandbox/testdir1', 'sandbox/testdir2')
		dir2Files = subprocess.check_output(['ls', 'sandbox/testdir2']).split()
		self.assertEqual(len(dir2Files), 3)
		self.assertTrue('file1' in dir2Files)
		self.assertTrue('file2' in dir2Files)
		self.assertTrue('file3' in dir2Files)

	def genArgs(self):
		args = cOS.genArgs({'k1': 'v1', 'k2' : 'v2', 'k3' : 'v3'})
		self.assertEqual(args, '-k3 v3 -k2 v2 -k1 v1')

	def getFrameRange(self):
		info = cOS.getFrameRange('sandbox/seq/frame.%04d.exr')
		self.assertEqual(info['min'], 1510)
		self.assertEqual(info['max'], 1519)

	def validateFrameFile(self):
		frameText = cOS.getFirstFileFromFrameRangeText('sandbox/seq/frame.%04d.exr 1510-1519')
		self.assertEqual(frameText, 'sandbox/seq/frame.1510.exr')

		frameText = cOS.getFirstFileFromFrameRangeText('sandbox/seq/newFrame.%04d.exr')
		self.assertEqual(frameText, 'sandbox/seq/newFrame.0001.exr')

		frameText = cOS.getFirstFileFromFrameRangeText('sandbox/seq/frame.1510.exr')
		self.assertEqual(frameText, 'sandbox/seq/frame.1510.exr')

		frameText = cOS.getFirstFileFromFrameRangeText('sandbox/seq/frame.exr')
		self.assertEqual(frameText, False)


	def removeStartingSlash(self):
		res = cOS.removeStartingSlash('/path/to/file')
		self.assertEqual(res, 'path/to/file')

	def normalizeDir(self):
		res = cOS.normalizeDir('/path\\to/file')
		self.assertEqual(res, '/path/to/file/')

	def normalizeExtension(self):
		norm = cOS.normalizeExtension('some/path.ABC')
		self.assertEqual(norm, 'some/path.abc')
		norm = cOS.normalizeExtension('some/path.abc')
		self.assertEqual(norm, 'some/path.abc')
		norm = cOS.normalizeExtension('Some/Path with/spaces.ABC')
		self.assertEqual(norm, 'Some/Path with/spaces.abc')

	def upADir(self):
		parent = cOS.upADir('path/to/a/file/')
		self.assertEqual(parent, 'path/to/a/')
		parent = cOS.upADir('path/to/a/file.txt')
		self.assertEqual(parent, 'path/to/')

	def join(self):
		joined = cOS.join('/path/to/a/directory/', '/path/to/a/file.txt')
		self.assertEqual(joined, '/path/to/a/directory/path/to/a/file.txt')

	def getFiles(self):
		# fix: this test is pretty bad :\
		root = os.path.abspath(
			os.path.join(
				os.path.dirname(os.path.realpath(__file__)),
				'../')
			)
		print root

		print 'test one:'
		files = cOS.getFiles(root,
			fileIncludes='*.py',
			folderExcludes=['.git','node_modules'],
			filesOnly=True

		)
		print '\n'.join(files)
		self.assertTrue(len(files) > 4)
		self.assertTrue(len(files) < 10)

		print '\ntest two:'
		files = cOS.getFiles(root,
			fileIncludes=['__init__.py'],
			fileExcludes=['*'],
			filesOnly=True,
		)
		print '\n'.join(files)
		self.assertTrue(len(files) == 4)

		print '\ntest three:'
		files = cOS.getFiles(root,
			fileExcludes=['.*'],
			folderExcludes=['.*', 'node_modules'],
			filesOnly=True,
		)
		print '\n'.join(files)
		self.assertTrue(len(files) > 4)
		self.assertTrue(len(files) < 40)

		print '\ntest four:'
		files = cOS.getFiles(root,
			fileIncludes=[
				'*cOS/cOS/cOS.py',
				'*cOS/cOS/__init__.py',
			],
			fileExcludes=['*.py', '.git*'],
			folderExcludes=['.git','node_modules'],
			includeAfterExclude=True,
			filesOnly=True,
		)
		print '\n'.join(files)
		self.assertTrue(len(files) > 4)
		self.assertTrue(len(files) < 40)

	def removeFile(self):
		self.assertTrue(os.path.isfile('sandbox/file.mb'))
		cOS.removeFile('sandbox/file.mb')
		self.assertTrue(not os.path.isfile('sandbox/file.mb'))
		ret = cOS.removeFile('sandbox/file.mb')
		self.assertTrue(ret != True)

	def removeDir(self):
		self.assertTrue(os.path.isdir('sandbox/emptyDir'))
		cOS.removeDir('sandbox/emptyDir')
		self.assertTrue(not os.path.isdir('sandbox/emptyDir'))
		ret = cOS.removeDir('sandbox/emptyDir')
		self.assertTrue(ret != True)

	# fix: can't really test this as we don't know what the
	# directory should be
	# def cwd(self):
	# 	cwd = cOS.cwd()
	# 	self.assertTrue()

	def ensureArray(self):
		self.assertEqual(cOS.ensureArray([1,2,3]), [1,2,3])
		self.assertEqual(cOS.ensureArray('abc'), ['abc'])
		self.assertEqual(cOS.ensureArray(None), [])
		self.assertEqual(cOS.ensureArray((1,2,3)), [1,2,3])

	def collectFiles(self):
		os.system('rm -rf seq')
		files = cOS.collectFiles('sandbox', 'mb', '')
		self.assertEqual(sorted(files), sorted([cOS.getPathInfo(f) for f in ['sandbox/file_v001.mb', 'sandbox/file.mb']]))
		files = cOS.collectFiles('sandbox', 'mb', 'sandbox/file_v001.mb')
		self.assertEqual(sorted(files), sorted([cOS.getPathInfo(f) for f in ['sandbox/file.mb']]))

	def collectAllFiles(self):
		files = cOS.collectAllFiles('sandbox/testdir2')
		self.assertEqual(sorted(files), sorted([cOS.getPathInfo(f) for f in ['sandbox/testdir2/file1']]))

	def isWindows(self):
		self.assertTrue(cOS.isWindows())

	def getCommandOutput(self):
		out, err = cOS.getCommandOutput('jkfsdajkl')
		self.assertTrue(out == False)
		self.assertTrue(err)

		testFile = cOS.getDirName(os.path.realpath(__file__)) + \
			'testOutput/simple.py'
		out, err = cOS.getCommandOutput('python ' + testFile)
		print 'out:', out
		print 'err:', err
		self.assertTrue('hello world' in out)
		self.assertEqual(err, False)

	def normalizeFramePadding(self):
		self.assertEqual(cOS.normalizeFramePadding('C:/Trash/abc.####.png'), 'C:/Trash/abc.%04d.png')
		self.assertEqual(cOS.normalizeFramePadding('C:/Trash/abc.$F6.png'), 'C:/Trash/abc.%06d.png')
		self.assertEqual(cOS.normalizeFramePadding('C:/Trash/abc.%04d.png'), 'C:/Trash/abc.%04d.png')
		self.assertEqual(cOS.normalizeFramePadding('C:/Trash/abc.$F.png'), 'C:/Trash/abc.%d.png')
		self.assertEqual(cOS.normalizeFramePadding('C:/Trash/abc.21.png'), 'C:/Trash/abc.%d.png')
		self.assertEqual(cOS.normalizeFramePadding('C:/Trash/abc.p3q0#$93bhn.png'), 'C:/Trash/abc.p3q0#$93bhn.png')
		self.assertEqual(cOS.normalizeFramePadding('C:/Trash/abc.png'), 'C:/Trash/abc.png')

if __name__ == '__main__':
	tryout.run(test)
